queue_job_batch/models/queue_job_batch.py

```diff
@@ -52,72 +54,75 @@ class QueueJobBatch(models.Model):
         "res.company",
         readonly=True,
     )
-    is_read = fields.Boolean(default=True)
+    is_read = fields.Boolean()
     completeness = fields.Float(
         compute="_compute_job_count",
     )
     failed_percentage = fields.Float(
         compute="_compute_job_count",
     )
 
-    def enqueue(self):
-        self.filtered(lambda r: r.state == "draft").write({"state": "enqueued"})
-        for record in self:
-            record.check_state()
+    def _get_state(self):
+        self.ensure_one()
+        job_states = set(self.job_ids.grouped("state").keys())
+        if all(state in ("done", "cancelled", "failed") for state in job_states):
+            return "finished"
+        elif {"done", "started"} & job_states:
+            return "progress"
+        elif "enqueued" in job_states:
+            return "enqueued"
+        return "pending"
 
     def check_state(self):
-        self.ensure_one()
-        if self.state == "enqueued" and any(
-            job.state not in ["pending", "enqueued"] for job in self.job_ids
-        ):
-            self.write({"state": "progress"})
-        if self.state != "progress":
-            return True
-        if all(job.state == "done" for job in self.job_ids):
-            self.write(
-                {
-                    "state": "finished",
-                    "is_read": False,
-                }
-            )
-        return True
+        for rec in self:
+            if (state := rec._get_state()) != rec.state:
+                rec.state = state
 
     def set_read(self):
-        res = self.write({"is_read": True})
-        notifications = []
-        channel = "queue.job.batch/updated"
-        notifications.append([self.env.user.partner_id, channel, {}])
-        self.env["bus.bus"]._sendmany(notifications)
-        return res
+        for rec in self:
+            if rec.is_read or rec.state != "finished":
+                continue
+            rec.is_read = True
+            rec.user_id._bus_send("queue.job.batch/updated", {"batch_read": True})
 
     @api.model
     def get_new_batch(self, name, **kwargs):
         vals = kwargs.copy()
-        company_id = self.env.user.company_id.id
-
-        if "company_id" in self.env.context:
-            company_id = self.env.context["company_id"]
-
         vals.update(
             {
                 "user_id": self.env.uid,
                 "name": name,
-                "state": "draft",
-                "company_id": company_id,
+                "company_id": self.env.company.id or self.env.user.company_id.id,
             }
         )
-        return self.sudo().create(vals).with_user(self.env.uid)
+        record = self.sudo().create(vals).with_user(self.env.uid)
+        record.user_id._bus_send("queue.job.batch/updated", {"batch_created": True})
+        return record
 
-    @api.depends("job_ids")
+    @api.depends("job_ids.state")
```

Comment:
do we really need this depends? Job count is not stored.

Reply:
Not really, for the UI. I'd be necessary when reading the field programatically, though, without invalidating caches. It's a habit 😓